src/AeroSharp/DataAccess/MapAccess/Parsers/MapBinParser.cs

```diff
@@ -0,0 +1,63 @@
+﻿using AeroSharp.DataAccess.Exceptions;
+using Aerospike.Client;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace AeroSharp.DataAccess.MapAccess.Parsers;
+
+/// <inheritdoc cref="IMapBinParser"/>"/>
+internal sealed class MapBinParser : IMapBinParser
+{
+    public KeyValuePair<object, object> ParseOne(Record record, string bin)
+    {
+        ValidateBinAndThrow(record, bin);
+
+        var binValues = ValidateAndGetBinValues(record, bin);
+
+        return ValidateAndGetFirstKeyValuePair(binValues, bin);
+    }
+
+    private static void ValidateBinAndThrow(Record record, string bin)
+    {
+        if (!record?.bins?.ContainsKey(bin) ?? true)
+        {
+            throw new BinNotFoundException($"Map record does not contain bin \"{bin}\".");
+        }
+
+        if (record.bins[bin] is null)
+        {
+            throw new MapEntryNotFoundException($"Map entry not found in bin \"{bin}\".");
+        }
+    }
+
+    private static IEnumerable<object> ValidateAndGetBinValues(Record record, string bin)
+    {
+        if (!record.bins.TryGetValue(bin, out var binObj) || binObj is not List<object> binValues)
+        {
+            throw new UnexpectedDataFormatException(
+                $"Unable to parse map entry from bin \"{bin}\"."
+            );
+        }
+
+        if (!binValues.Any())
```

Comment:
So I pulled this into SharpLab, comparing `Any` vs `Count` vs pattern matching:

<img width="1411" alt="image" src="https://github.com/wayfair-incubator/AeroSharp/assets/5827236/b2ff7651-33fa-4a0f-8d6e-506aa13ac120">

`Any` relies on Linq which is known to cause allocations, which we might want to avoid on a hot code path-- which this likely is. Comparing count against 0 has no allocations and communicates similarly the intent.

Pattern matching actually does the same thing, but also guarantees not being null, which you handled on line 35.

I imagine we're using Any in a number of places throughout this library-- so we could opt to continue for consistency. It might be another opportunity for a new issue to address this in the codebase as a whole.

Reply:
I vote for starting this as the "new way" and if @farmerau doesn't mind opening an issue for the rest to replace any .Any's lingering around?

Reply:
Nice find! I'm also a fan of pattern matching due to its expressiveness. I'll swap to that here.

+1 to @awesor's suggestion that we open up an issue for identifying and replacing other usages of `Any`.

Reply:
Yep, I'll open up an issue for it.

Reply:
Scratch that. Since we're multi-targeting and supporting .NET 6.0, which only supports up to C# 10, we can't use that form of pattern matching since it's a C# 11 feature. We'd need to do something funky like this, which I don't love:

```csharp
#if NET6_0
        if (binValues.Count == 0)
        {
            throw new MapEntryNotFoundException($"Map entry not found in bin \"{bin}\".");
        }
#endif

#if NET7_0
        if (binValues is [])
        {
            throw new MapEntryNotFoundException($"Map entry not found in bin \"{bin}\".");
        }
#endif
```

That said, I'm going to use the count check.

Reply:
Actually, [a quick search](https://github.com/search?q=repo%3Awayfair-incubator%2FAeroSharp%20.Any()&type=code) reveals that the only times we're using `Any` are on `IEnumerable<T>`. 

It's incredibly arguable whether or not the collections _should be_ `IEnumerable` and not `IList` (where we could perform such optimization) but that's another problem for another day.

Reply:
> Scratch that. Since we're multi-targeting and supporting .NET 6.0, which only supports up to C# 10, we can't use that form of pattern matching since it's a C# 11 feature. We'd need to do something funky like this, which I don't love
> That said, I'm going to use the count check.

Ah, bummer! Count makes sense, though. :)